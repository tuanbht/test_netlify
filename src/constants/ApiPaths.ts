export const API_URL = process.env.REACT_APP_BASE_API;

export const ORDER_DETAILS_PATH = '/orders/:id/payment/';
export const MARK_AS_PAID_PATH = '/orders/:id/mask_as_paid/';
export const CANCEL_ORDER_PATH = '/orders/:id/payment/cancel/';
