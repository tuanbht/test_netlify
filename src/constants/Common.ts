export const VIEW_CRYPTO_TRANSACTION = process.env.REACT_APP_VIEW_CRYPTO_TRANSACTION;
