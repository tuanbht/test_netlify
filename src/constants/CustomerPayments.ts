import React from 'react';
import { ReactComponent as IconBTC } from '../assets/images/icons/icon-btc.svg';
import { ReactComponent as IconETH } from '../assets/images/icons/icon-eth.svg';
import { ReactComponent as IconUSDT } from '../assets/images/icons/icon-usdt.svg';

export const STEPS = {
  CRYPTO: 0,
  WALLET: 1,
  PROCESSING: 2,
  COMPLETE: 3,
};

export interface CryptoCurrencies {
  prefix: string;
  shortName: string;
  fullName: string;
  logo: React.FC<React.SVGProps<SVGSVGElement>>;
  walletAddress: string;
  amount: number;

  setWalletAddress(walletAddress: string): void;
  setAmount(amount: number): void;
}

export class CryptoCurrencies implements CryptoCurrencies {
  static initWithData(
    prefix: string,
    shortName: string,
    fullName: string,
    logo: React.FC<React.SVGProps<SVGSVGElement>>,
  ): CryptoCurrencies {
    const crypto = new CryptoCurrencies();
    crypto.prefix = prefix;
    crypto.shortName = shortName;
    crypto.fullName = fullName;
    crypto.logo = logo;
    return crypto;
  }
  setWalletAddress(walletAddress: string): void {
    this.walletAddress = walletAddress;
  }
  setAmount(amount: number): void {
    this.amount = amount;
  }
}

export const CRYPTO_CURRENCIES = {
  BITCOIN: CryptoCurrencies.initWithData('bitcoin', 'BTC', 'Bitcoin', IconBTC),
  ETHEREUM: CryptoCurrencies.initWithData('ethereum', 'ETH', 'Ethereum', IconETH),
  USDT: CryptoCurrencies.initWithData('usdt', 'USDT', 'USD Tether', IconUSDT),
};
