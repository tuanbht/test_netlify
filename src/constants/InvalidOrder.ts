export const ORDER_CANCELLED_CONTENT = {
  title: 'Order cancelled',
  subtitle: 'This order has been cancelled and is no longer available. To place a new order, contact your Merchant.',
};

export const ORDER_EXPIRED_CONTENT = {
  title: 'Order expired',
  subtitle: 'This order expired and is no longer available. To place a new order, contact your Merchant.',
};
