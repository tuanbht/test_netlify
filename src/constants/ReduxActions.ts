const SUCCESS_SUFFIX = '_SUCCESS';
const FAILURE_SUFFIX = '_FAIL';

export const GET_ORDER_DETAILS = 'GET_ORDER_DETAILS';
export const MARK_AS_PAID_ORDER = 'MARK_AS_PAID_ORDER';
export const CANCEL_ORDER = 'CANCEL_ORDER';

export const SAVE_CREDENTIAL = 'SAVE_CREDENTIAL';

export const ActionSuccessfully = (action: string): string => {
  return action.concat(SUCCESS_SUFFIX);
};

export const ActionFailure = (action: string): string => {
  return action.concat(FAILURE_SUFFIX);
};
