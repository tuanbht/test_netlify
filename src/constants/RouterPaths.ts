export const ORDER_PATH = '/:orderId/:token';
export const ORDER_EXPIRED_PATH = '/order-expired';
export const ORDER_CANCELLED_PATH = '/order-cancelled';
export const NOT_FOUND_PATH = '/';
