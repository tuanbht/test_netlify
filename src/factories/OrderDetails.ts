import { ORDER_STATUS, OrderDetails } from '../constants/CustomerPayments';
import faker from 'faker';

export const HardCodedOrderDetails = {
  status: 'order-status',
  orderNumber: 1,
  price: 1.0,
  storeName: 'store-name',
  storePhoneNumber: 'store-phone-number',
  markAsPaid: true,
  markAsPaidTime: 1601607445,
} as OrderDetails;

export const OrderDetailsEntity = {
  status: ORDER_STATUS.initialized,
  orderNumber: faker.random.number(),
  price: faker.random.number(),
  storeName: faker.company.companyName(),
  storePhoneNumber: faker.phone.phoneNumber(),
  markAsPaid: true,
  markAsPaidTime: new Date().getTime() / 1000,
} as OrderDetails;

export const buildOrderDetailsResponse = (props: {
  status?: string;
  hasStoreDetails?: boolean;
  hasCryptoETH?: boolean;
  hasCryptoBTC?: boolean;
  hasCryptoUSDT?: boolean;
  markAsPaid?: boolean;
}): any => ({
  data: {
    type: 'orders',
    attributes: {
      'external-order-id': 'NjKvLP1o',
      price: '67.77',
      'phone-number': '1-924-340-4717',
      name: 'Peggy Nikolaus',
      status: props.status,
      'marked-as-paid': props.markAsPaid,
      'marked-as-paid-at': props.markAsPaid && 1601607445,
    },
    relationships: {
      store: {
        data: {
          type: 'stores',
          id: '2',
        },
      },
      payments: {
        data: [
          {
            type: 'payments',
            id: '3',
          },
          {
            type: 'payments',
            id: '4',
          },
        ],
      },
    },
  },
  included: [
    props.hasStoreDetails && {
      id: '2',
      type: 'stores',
      links: {
        self: 'http://192.168.0.80:3000/v1/stores/2',
      },
      attributes: {
        name: 'Marilu Harvey MD',
        address: 'Suite 612 42874 Timmy Drive, Kerlukechester, IN 43320',
        'phone-number': '993.885.9672',
      },
      relationships: {
        merchant: {
          links: {
            self: 'http://192.168.0.80:3000/v1/stores/2/relationships/merchant',
            related: 'http://192.168.0.80:3000/v1/stores/2/merchant',
          },
        },
        orders: {
          links: {
            self: 'http://192.168.0.80:3000/v1/stores/2/relationships/orders',
            related: 'http://192.168.0.80:3000/v1/stores/2/orders',
          },
          data: [
            {
              type: 'orders',
              id: '2',
            },
          ],
        },
      },
    },
    props.hasCryptoETH && {
      id: '3',
      type: 'payments',
      attributes: {
        amount: '0.47646764203043',
        'destination-wallet': '0x3f5CE5FBFe3E9af3971dD833D26bA9b5C936f0bE',
        'payment-type': 'ETH',
      },
    },
    props.hasCryptoBTC && {
      id: '4',
      type: 'payments',
      attributes: {
        amount: '0.755828606587089',
        'destination-wallet': '1A1zP1eP5QGefi2DMPTfTL5SLmv7DivfNa',
        'payment-type': 'BTC',
      },
    },
    props.hasCryptoUSDT && {
      id: '5',
      type: 'payments',
      attributes: {
        amount: '0.755828606587089',
        'destination-wallet': '1A1zP1eP5QGefi2DMPTfTL5SLmv7DivfNa',
        'payment-type': 'USDT',
      },
    },
  ],
});
