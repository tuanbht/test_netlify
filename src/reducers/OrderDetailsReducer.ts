import { createReducer } from '@reduxjs/toolkit';
import { CRYPTO_CURRENCIES, ORDER_STATUS, OrderDetails as OrderDetailsModel } from '../constants/CustomerPayments';
import { ActionFailure, ActionSuccessfully, GET_ORDER_DETAILS } from '../constants/ReduxActions';
import { NOT_FOUND_PATH, ORDER_CANCELLED_PATH, ORDER_EXPIRED_PATH } from '../constants/RouterPaths';
import { find, get } from 'lodash';

const OrderDetails = createReducer(new OrderDetailsModel(), {
  [ActionSuccessfully(GET_ORDER_DETAILS)]: (state, action) => {
    const { data } = action.payload;
    const orderDetails: OrderDetailsModel = new OrderDetailsModel();

    orderDetails.status = get(data.data.attributes, 'status', ORDER_STATUS.undefined);

    if (orderDetails.status === ORDER_STATUS.cancelled) {
      window.location.assign(ORDER_CANCELLED_PATH);
      return state;
    }
    if (orderDetails.status === ORDER_STATUS.expired) {
      window.location.assign(ORDER_EXPIRED_PATH);
      return state;
    }

    orderDetails.orderNumber = get(data.data.attributes, 'external-order-id', 0);
    orderDetails.markAsPaid = get(data.data.attributes, 'marked-as-paid', false);
    orderDetails.markAsPaidTime = get(data.data.attributes, 'marked-as-paid-at', '');
    orderDetails.price = get(data.data.attributes, 'price', 0);

    const storeDetails = find(data.included, { type: 'stores' });
    if (storeDetails) {
      orderDetails.storeName = get(storeDetails.attributes, 'name', '');
      orderDetails.storePhoneNumber = get(storeDetails.attributes, 'phone-number', '');
    }

    const cryptoETH = find(data.included, {
      type: 'payments',
      attributes: { 'payment-type': CRYPTO_CURRENCIES.ETHEREUM.shortName },
    });
    const cryptoBTC = find(data.included, {
      type: 'payments',
      attributes: { 'payment-type': CRYPTO_CURRENCIES.BITCOIN.shortName },
    });
    const cryptoUSDT = find(data.included, {
      type: 'payments',
      attributes: { 'payment-type': CRYPTO_CURRENCIES.USDT.shortName },
    });

    if (cryptoETH) {
      CRYPTO_CURRENCIES.ETHEREUM.setAmount(get(cryptoETH.attributes, 'amount', 0));
      CRYPTO_CURRENCIES.ETHEREUM.setWalletAddress(get(cryptoETH.attributes, 'destination-wallet', ''));
    }
    if (cryptoBTC) {
      CRYPTO_CURRENCIES.BITCOIN.setAmount(get(cryptoBTC.attributes, 'amount', 0));
      CRYPTO_CURRENCIES.BITCOIN.setWalletAddress(get(cryptoBTC.attributes, 'destination-wallet', ''));
    }
    if (cryptoUSDT) {
      CRYPTO_CURRENCIES.USDT.setAmount(get(cryptoUSDT.attributes, 'amount', 0));
      CRYPTO_CURRENCIES.USDT.setWalletAddress(get(cryptoUSDT.attributes, 'destination-wallet', ''));
    }

    return orderDetails;
  },
  [ActionFailure(GET_ORDER_DETAILS)]: (state) => {
    window.location.assign(NOT_FOUND_PATH);
    return state;
  },
});

export default OrderDetails;
