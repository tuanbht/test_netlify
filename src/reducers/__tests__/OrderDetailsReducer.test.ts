import { ActionSuccessfully, GET_ORDER_DETAILS } from 'constants/ReduxActions';
import { AnyAction } from '@reduxjs/toolkit';
import { CRYPTO_CURRENCIES, ORDER_STATUS, OrderDetails } from 'constants/CustomerPayments';
import OrderDetailsReducer from '../OrderDetailsReducer';
import { ORDER_CANCELLED_PATH, ORDER_EXPIRED_PATH } from 'constants/RouterPaths';
import { buildOrderDetailsResponse } from 'factories/OrderDetails';

describe('OrderDetails', () => {
  beforeEach(() => {
    jest.spyOn(window.location, 'assign').mockReturnValue();
  });

  describe('GET_ORDER_DETAILS_SUCCESS', () => {
    let action: AnyAction;

    describe('order status is cancelled', () => {
      beforeEach(() => {
        action = {
          type: ActionSuccessfully(GET_ORDER_DETAILS),
          payload: {
            data: buildOrderDetailsResponse({ status: ORDER_STATUS.cancelled }),
          },
        };
        OrderDetailsReducer(new OrderDetails(), action);
      });

      it('redirects to cancelled page', () => {
        expect(window.location.assign).toBeCalledWith(ORDER_CANCELLED_PATH);
      });
    });

    describe('order status is expired', () => {
      beforeEach(() => {
        action = {
          type: ActionSuccessfully(GET_ORDER_DETAILS),
          payload: {
            data: buildOrderDetailsResponse({ status: ORDER_STATUS.expired }),
          },
        };
        OrderDetailsReducer(new OrderDetails(), action);
      });

      it('redirects to expired page', () => {
        expect(window.location.assign).toBeCalledWith(ORDER_EXPIRED_PATH);
      });
    });

    describe('order has expired time', () => {
      beforeEach(() => {
        jest.useFakeTimers();
      });

      describe('expired time is valid', () => {
        beforeEach(() => {
          action = {
            type: ActionSuccessfully(GET_ORDER_DETAILS),
            payload: {
              data: buildOrderDetailsResponse({ isExpired: false }),
            },
          };
          OrderDetailsReducer(new OrderDetails(), action);
        });

        it('redirects to expired page when expired time come', () => {
          jest.runAllTimers();
          expect(window.location.assign).toBeCalledWith(ORDER_EXPIRED_PATH);
        });
      });
    });

    describe('does not have store details and crypto payment details', () => {
      beforeEach(() => {
        action = {
          type: ActionSuccessfully(GET_ORDER_DETAILS),
          payload: {
            data: buildOrderDetailsResponse({ status: ORDER_STATUS.initialized }),
          },
        };
      });

      it('returns orderDetails without store details and crypto payment details', () => {
        expect(OrderDetailsReducer(new OrderDetails(), action)).toMatchSnapshot();
        expect(CRYPTO_CURRENCIES).toMatchSnapshot();
      });
    });

    describe('has store details', () => {
      beforeEach(() => {
        action = {
          type: ActionSuccessfully(GET_ORDER_DETAILS),
          payload: {
            data: buildOrderDetailsResponse({ hasStoreDetails: true }),
          },
        };
      });

      it('returns orderDetails with store details', () => {
        expect(OrderDetailsReducer(new OrderDetails(), action)).toMatchSnapshot();
      });
    });

    describe('has crypto ETH details', () => {
      beforeEach(() => {
        action = {
          type: ActionSuccessfully(GET_ORDER_DETAILS),
          payload: {
            data: buildOrderDetailsResponse({ hasCryptoETH: true }),
          },
        };
        OrderDetailsReducer(new OrderDetails(), action);
      });

      it('returns crypto ETH details', () => {
        expect(CRYPTO_CURRENCIES).toMatchSnapshot();
      });
    });

    describe('has crypto USDT details', () => {
      beforeEach(() => {
        action = {
          type: ActionSuccessfully(GET_ORDER_DETAILS),
          payload: {
            data: buildOrderDetailsResponse({ hasCryptoUSDT: true }),
          },
        };
        OrderDetailsReducer(new OrderDetails(), action);
      });

      it('returns crypto USDT details', () => {
        expect(CRYPTO_CURRENCIES).toMatchSnapshot();
      });
    });

    describe('transaction hash', () => {
      describe('has none', () => {
        beforeEach(() => {
          action = {
            type: ActionSuccessfully(GET_ORDER_DETAILS),
            payload: {
              data: buildOrderDetailsResponse({
                hasCryptoETH: true,
                hasCryptoUSDT: true,
              }),
            },
          };
        });

        it('returns order details with empty transaction hash', () => {
          expect(OrderDetailsReducer(new OrderDetails(), action)).toMatchSnapshot();
        });
      });

      describe('has one', () => {
        describe('transaction hash ETH', () => {
          beforeEach(() => {
            action = {
              type: ActionSuccessfully(GET_ORDER_DETAILS),
              payload: {
                data: buildOrderDetailsResponse({ hasCryptoETH: true, hasHashETH: true }),
              },
            };
          });

          it('returns order details with transaction hash ETH', () => {
            expect(OrderDetailsReducer(new OrderDetails(), action)).toMatchSnapshot();
          });
        });

        describe('transaction hash USDT', () => {
          beforeEach(() => {
            action = {
              type: ActionSuccessfully(GET_ORDER_DETAILS),
              payload: {
                data: buildOrderDetailsResponse({ hasCryptoUSDT: true, hasHashUSDT: true }),
              },
            };
          });

          it('returns order details with transaction hash USDT', () => {
            expect(OrderDetailsReducer(new OrderDetails(), action)).toMatchSnapshot();
          });
        });
      });

      describe('has many', () => {
        beforeEach(() => {
          action = {
            type: ActionSuccessfully(GET_ORDER_DETAILS),
            payload: {
              data: buildOrderDetailsResponse({
                hasCryptoETH: true,
                hasHashETH: true,
                hasCryptoUSDT: true,
                hasHashUSDT: true,
              }),
            },
          };
        });

        it('returns order details with the first transaction hash', () => {
          expect(OrderDetailsReducer(new OrderDetails(), action)).toMatchSnapshot();
        });
      });
    });
  });
});
