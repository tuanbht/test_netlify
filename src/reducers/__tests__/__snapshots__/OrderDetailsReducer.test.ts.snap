// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderDetails GET_ORDER_DETAILS_SUCCESS does not have store details and crypto payment details returns orderDetails without store details and crypto payment details 1`] = `
OrderDetails {
  "isEmpty": [Function],
  "markAsPaid": false,
  "markAsPaidTime": "",
  "orderNumber": "NjKvLP1o",
  "price": "67.77",
  "status": "initialized",
  "storeName": "",
  "storePhoneNumber": "",
}
`;

exports[`OrderDetails GET_ORDER_DETAILS_SUCCESS does not have store details and crypto payment details returns orderDetails without store details and crypto payment details 2`] = `
Object {
  "BITCOIN": CryptoCurrencies {
    "amount": 0,
    "fullName": "Bitcoin",
    "logo": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
    "prefix": "bitcoin",
    "shortName": "BTC",
    "walletAddress": "",
  },
  "ETHEREUM": CryptoCurrencies {
    "amount": 0,
    "fullName": "Ethereum",
    "logo": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
    "prefix": "ethereum",
    "shortName": "ETH",
    "walletAddress": "",
  },
  "USDT": CryptoCurrencies {
    "amount": 0,
    "fullName": "USD Tether",
    "logo": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
    "prefix": "usdt",
    "shortName": "USDT",
    "walletAddress": "",
  },
}
`;

exports[`OrderDetails GET_ORDER_DETAILS_SUCCESS has crypto BTC details returns crypto BTC details 1`] = `
Object {
  "BITCOIN": CryptoCurrencies {
    "amount": "0.755828606587089",
    "fullName": "Bitcoin",
    "logo": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
    "prefix": "bitcoin",
    "shortName": "BTC",
    "walletAddress": "1A1zP1eP5QGefi2DMPTfTL5SLmv7DivfNa",
  },
  "ETHEREUM": CryptoCurrencies {
    "amount": "0.47646764203043",
    "fullName": "Ethereum",
    "logo": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
    "prefix": "ethereum",
    "shortName": "ETH",
    "walletAddress": "0x3f5CE5FBFe3E9af3971dD833D26bA9b5C936f0bE",
  },
  "USDT": CryptoCurrencies {
    "amount": 0,
    "fullName": "USD Tether",
    "logo": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
    "prefix": "usdt",
    "shortName": "USDT",
    "walletAddress": "",
  },
}
`;

exports[`OrderDetails GET_ORDER_DETAILS_SUCCESS has crypto ETH details returns crypto ETH details 1`] = `
Object {
  "BITCOIN": CryptoCurrencies {
    "amount": 0,
    "fullName": "Bitcoin",
    "logo": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
    "prefix": "bitcoin",
    "shortName": "BTC",
    "walletAddress": "",
  },
  "ETHEREUM": CryptoCurrencies {
    "amount": "0.47646764203043",
    "fullName": "Ethereum",
    "logo": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
    "prefix": "ethereum",
    "shortName": "ETH",
    "walletAddress": "0x3f5CE5FBFe3E9af3971dD833D26bA9b5C936f0bE",
  },
  "USDT": CryptoCurrencies {
    "amount": 0,
    "fullName": "USD Tether",
    "logo": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
    "prefix": "usdt",
    "shortName": "USDT",
    "walletAddress": "",
  },
}
`;

exports[`OrderDetails GET_ORDER_DETAILS_SUCCESS has crypto USDT details returns crypto USDT details 1`] = `
Object {
  "BITCOIN": CryptoCurrencies {
    "amount": "0.755828606587089",
    "fullName": "Bitcoin",
    "logo": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
    "prefix": "bitcoin",
    "shortName": "BTC",
    "walletAddress": "1A1zP1eP5QGefi2DMPTfTL5SLmv7DivfNa",
  },
  "ETHEREUM": CryptoCurrencies {
    "amount": "0.47646764203043",
    "fullName": "Ethereum",
    "logo": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
    "prefix": "ethereum",
    "shortName": "ETH",
    "walletAddress": "0x3f5CE5FBFe3E9af3971dD833D26bA9b5C936f0bE",
  },
  "USDT": CryptoCurrencies {
    "amount": "0.755828606587089",
    "fullName": "USD Tether",
    "logo": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
    "prefix": "usdt",
    "shortName": "USDT",
    "walletAddress": "1A1zP1eP5QGefi2DMPTfTL5SLmv7DivfNa",
  },
}
`;

exports[`OrderDetails GET_ORDER_DETAILS_SUCCESS has store details returns orderDetails with store details 1`] = `
OrderDetails {
  "isEmpty": [Function],
  "markAsPaid": false,
  "markAsPaidTime": "",
  "orderNumber": "NjKvLP1o",
  "price": "67.77",
  "status": "undefined",
  "storeName": "Marilu Harvey MD",
  "storePhoneNumber": "993.885.9672",
}
`;
